pkg/streamingpromql/operators/instant_vector_selector.go

```diff
@@ -84,6 +105,20 @@ func (v *InstantVectorSelector) NextSeries(ctx context.Context) (types.InstantVe
 			if !ok || t < ts-v.Selector.LookbackDelta.Milliseconds() {
 				continue
 			}
+			if h != nil {
+				if t == lastHistogramT {
+					// Reuse exactly the same FloatHistogram as last time.
+					// PeekPrev can return a new FloatHistogram instance with the same underlying bucket slices as a previous call
```

Comment:
`MemoizedSeriesIterator` calls `AtFloatHistogram(nil)` on its inner iterator when `Seek` or `Next` is called to store the previous point for lookback. We also call `AtFloatHistogram(nil)` ourselves to retrieve the current point above.

However, the inner iterator (which is always a `floatHistogramIterator`, or eventually calls into `floatHistogramIterator`) will always return a new `FloatHistogram` instance with a reference to the same bucket and span slices ([source](https://github.com/prometheus/prometheus/blob/52a73ec87166163c809cef984dd51f3f56c5d88e/tsdb/chunkenc/float_histogram.go#L773-L776)).

For example, imagine we are evaluating a range query with a 1m step and 5m lookback from t=5m to t=10m. The series has a histogram at t=5m.

At t=5m:
* We call `Seek(5m)`. 
* We then call `AtFloatHistogram(nil)` and get a `FloatHistogram` for t=5m (let's call it `fh1`)

At t=6m:
* We call `Seek(6m)`. 
  * As part of this, `Seek()` [calls](https://github.com/prometheus/prometheus/blob/52a73ec87166163c809cef984dd51f3f56c5d88e/storage/memoized_iterator.go#L95) `Next()` which stores the previous histogram as `prevFloatHistogram` by calling `AtFloatHistogram(nil)` [again](https://github.com/prometheus/prometheus/blob/52a73ec87166163c809cef984dd51f3f56c5d88e/storage/memoized_iterator.go#L116). This returns another `FloatHistogram` instance (let's call it `fh2`) with a reference to the [same bucket and span slices](https://github.com/prometheus/prometheus/blob/52a73ec87166163c809cef984dd51f3f56c5d88e/tsdb/chunkenc/float_histogram.go#L773-L776) as `fh1`.
  * `Seek()` then [returns](https://github.com/prometheus/prometheus/blob/52a73ec87166163c809cef984dd51f3f56c5d88e/storage/memoized_iterator.go#L101) `ValNone` because we've reached the end of the series. 
* We then call `PeekPrev()`, which returns `prevFloatHistogram`, which is `fh2`.

At t=7m to t=10m, the same thing happens: we eventually call `PeekPrev()`, which returns `fh2` exactly.

Reply:
Ah, I see. Thanks!